### PR TITLE
ref(prism): Dynamically import language files

### PR DESCRIPTION
**Pull request by vuluongj20**
Dynamically import Prism language files based on the `language` prop, based on @evanpurkhiser's suggestion.

**Before ——** `XQuery` block is not parsed
Screenshot 2023-02-03 at 4 42 52 PM

**After ——** `XQuery` language file is dynamically imported and code block is parsed
Screenshot 2023-02-03 at 4 43 13 PM

### CODE REVIEW

File: static/app/components/codeSnippet.tsx [hunk 1]
```diff
@@ -33,10 +29,27 @@ export function CodeSnippet({
...
+    import(`prismjs/components/prism-${language}.min`)
+      .then(() => Prism.highlightElement(element))
+      .catch(error => {
+        // eslint-disable-next-line no-console
+        console.warn(
+          'Prism language not found. Check the `language` prop passed to `<CodeSnippet />`.',
+          error.message
+        );
+      });
+  }, [children, language]);
```
Comment (evanpurkhiser) [inline static/app/components/codeSnippet.tsx:52]: You can use `await` with a try catch for this too to make it a bit cleaner.

File: static/app/utils/marked.tsx [hunk 2]
```diff
@@ -51,10 +51,28 @@ marked.setOptions({
...
+        callback?.(null, Prism.highlight(code, Prism.languages[lang], lang));
+      })
+      .catch(error => {
+        callback?.(error, code);
+        // eslint-disable-next-line no-console
+        console.warn(
+          'Prism language not found. Error encountered while trying to highlight markdown code block with marked.',
+          error.message
+        );
+      });
```
Comment (evanpurkhiser) [inline static/app/utils/marked.tsx:74]: Should we go ahead and put this in a util?

File: static/app/components/codeSnippet.tsx [hunk 3]
```diff
@@ -14,7 +10,7 @@ import space from 'sentry/styles/space';
 
 interface CodeSnippetProps {
   children: string;
-  language: keyof typeof Prism.languages;
+  language: string;
```
Comment (vuluongj20) [inline static/app/components/codeSnippet.tsx:13]: Unfortunately there doesn't seem to be any type declarations for available Prism languages, [language names are simply typed as `string`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/prismjs/components/index.d.ts)

Comment (evanpurkhiser) [reply static/app/components/codeSnippet.tsx:13]: Yeah that's unfortuante.

File: static/app/utils/marked.tsx (hunk 2 above)
Comment (vuluongj20) [reply static/app/utils/marked.tsx:74]: Wouldn't we still have to execute `callback?.(null, Prism.highlight(code, Prism.languages[lang], lang));` with a util? (maybe passed in as a callback once the file is loaded)

File: static/app/components/codeSnippet.tsx (hunk 1 above)
Comment (vuluongj20) [reply static/app/components/codeSnippet.tsx]: great suggestion! https://github.com/getsentry/sentry/pull/44146/commits/dd89ac012adc713434d7852d5d7ae74407227a4f

**Comment by vuluongj20** [comment]
@billyvg This addresses the concern you raised in https://github.com/getsentry/sentry/pull/43837#discussion_r1092607023

File: static/app/utils/marked.tsx (hunk 2 above)
Comment (evanpurkhiser) [reply static/app/utils/marked.tsx:74]: yeah exactly

Comment (vuluongj20) [reply static/app/utils/marked.tsx]: boom https://github.com/getsentry/sentry/pull/44146/commits/f5819e86abc734fe178cc9df7ed76972bba7f235

**Review by evanpurkhiser — APPROVED** [review]

**Comment by evanpurkhiser** [comment]
Let's keep it open @vuluongj20 ?

File: static/app/utils/loadPrismLanguage.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+      return [
+        [lang, lang], // map the full language name to itself
+        ...(Array.isArray(value.alias) // map aliases to full language name
+          ? value.alias.map(alias => [alias, lang])
+          : [[value.alias, lang]]),
+      ];
+    })
+    .flat(1)
+);
```
Comment (vuluongj20) [inline static/app/utils/loadPrismLanguage.ts:23]: This new mapping object serves 2 purposes: mapping aliases to full language names (so now we can pass aliases like `js` into `loadPrismLanguage` and it'll work), and checking to see if a language/alias is supported by Prism.

File: static/app/utils/loadPrismLanguage.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    if (!language) {
+      if (!suppressExistenceWarning) {
+        // eslint-disable-next-line no-console
+        console.warn(
+          `No Prism grammar file for \`${lang}\` exists. Check the \`lang\` argument passed to \`loadPrismLanguage()\`.`
+        );
+      }
+
+      return;
+    }
```
Comment (vuluongj20) [inline static/app/utils/loadPrismLanguage.ts:55]: We want to suppress these existence warnings in components like `ActivityNote`.

**Pull request merged**